tensorflow/c/experimental/gradients/math_grad.cc

```diff
@@ -31,23 +36,188 @@ class AddGradientFunction : public GradientFunction {
     TF_RETURN_IF_ERROR(ops::Identity(ctx->ctx, {grad_inputs[0]},
                                      absl::MakeSpan(identity_outputs),
-                                     "Identity0"));
+                                     name.c_str()));
     (*grad_outputs)[0] = identity_outputs[0];
+
+    name = "Identity_B_" + std::to_string(counter);
     TF_RETURN_IF_ERROR(ops::Identity(ctx->ctx, {grad_inputs[0]},
                                      absl::MakeSpan(identity_outputs),
-                                     "Identity1"));
+                                     name.c_str()));
     (*grad_outputs)[1] = identity_outputs[0];
+
+    counter += 1;
     return Status::OK();
   }
   ~AddGradientFunction() override {}
+
+ private:
+  long counter;
+};
+
+class MatMulGradientFunction : public GradientFunction {
+ public:
+  explicit MatMulGradientFunction(std::vector<AbstractTensorHandle*> f_inputs)
+      : forward_inputs(f_inputs) {}
+
+  Status Compute(Context* ctx,
+                 absl::Span<AbstractTensorHandle* const> grad_inputs,
+                 std::vector<AbstractTensorHandle*>* grad_outputs) override {
+    /* Given upstream grad U and a matmul op A*B, the gradients are:
+     *
+     *    dA = U * B.T
+     *    dB = A.T * U
+     *
+     *    where A.T means `transpose(A)`
+     */
+
+    AbstractTensorHandle* upstream_grad = grad_inputs[0];
+    grad_outputs->resize(2);
+    std::vector<AbstractTensorHandle*> matmul_outputs(1);
+
+    // Gradient for A
+    std::string name = "mm_A_" + std::to_string(counter);
+    TF_RETURN_IF_ERROR(MatMul(ctx->ctx, {upstream_grad, forward_inputs[1]},
+                              absl::MakeSpan(matmul_outputs), name.c_str(),
+                              /*transpose_a = */ false,
+                              /*transpose_b = */ true));
+
+    (*grad_outputs)[0] = matmul_outputs[0];
+
+    // Gradient for B
+    name = "mm_B_" + std::to_string(counter);
+    TF_RETURN_IF_ERROR(MatMul(ctx->ctx, {forward_inputs[0], upstream_grad},
+                              absl::MakeSpan(matmul_outputs), name.c_str(),
+                              /*transpose_a = */ true,
+                              /*transpose_b = */ false));
+
+    (*grad_outputs)[1] = matmul_outputs[0];
+
+    counter += 1;  // update counter for names
+    return Status::OK();
+  }
+  ~MatMulGradientFunction() override {}
+
+ private:
+  long counter;
+  std::vector<AbstractTensorHandle*> forward_inputs;
+};
+
+class ReluGradientFunction : public GradientFunction {
+ public:
+  explicit ReluGradientFunction(std::vector<AbstractTensorHandle*> f_inputs)
+      : forward_inputs(f_inputs) {}
+
+  Status Compute(Context* ctx,
+                 absl::Span<AbstractTensorHandle* const> grad_inputs,
+                 std::vector<AbstractTensorHandle*>* grad_outputs) override {
+    AbstractTensorHandle* upstream_grad = grad_inputs[0];
+    AbstractTensorHandle* input_features = forward_inputs[0];
```

Comment:
Unless I'm misunderstanding, I think the `ReluGrad` Op asks for `features` which is the input to the Relu Op

Reply:
Either is correct in this case since all we care about is the [sign](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/relu_op_functor.h;l=51-55;drc=791f4f2b773208b21ece71e1d37204a91720346d) but I think it would be nice to be consistent with the python gradient function which uses the output.

Reply:
SGTM, I'll change that